### PR TITLE
Add a test that checks dynamic imports

### LINKED ISSUE #6620: A sporadic import failure in the CI when adding a global hook
Prerequisite for the #6434 

When adding a hook execution into the api layer - a seemingly unrelated test started to fail with a failure to import a module stored in the `python` cache subfolder:
https://conan-ci.jfrog.info/blue/organizations/jenkins/ConanTestSuite/detail/PR-6434/1/pipeline

The test under question:
https://github.com/conan-io/conan/blob/develop/conans/test/functional/command/config_install_test.py#L187

I was able to reproduce the problem locally. Debugging shown that despite all files being identically present in both with and without the hook situations - the import is failing. Chipping the problem down shown that the problem is somehow related to the `_load_module_from_file` being present.

After some investigation, I discovered the https://docs.python.org/3/library/importlib.html#importlib.invalidate_caches - which sounds like the tool for the job since in tests we do append the `sys.path` with folders created within the same Python interpreter instance, which sounds pretty dynamic to me. 
Also, a nice write up on the problem: http://ballingt.com/import-invalidate-caches/

And it did fix the problem for the Python 3.8 interpreter. But it failed on Python 2.7 since it's not implemented in there and the Python 3.6, for some other weird reason. 

After additional investigation, I have found this Python issue: https://bugs.python.org/issue33169 

> Note that it's solved for Python 3.7+.

The most interesting to our use case note: 

> importlib.invalidate_caches() doesn't clear the negative cache in sys.path_importer_cache.

And indeed! At the moment when a sporadic import fails - the `path_import_cache` contains a `None` value for the temp folder key. Here is a description of how they have reproduced it:
https://github.com/python/mypy/pull/4811

In #6434 I copied the "fixed" cache invalidation from the 3.7+:
https://github.com/python/cpython/blob/521995205a2cb6b504fe0e39af22a81f785350a3/Lib/importlib/_bootstrap_external.py#L1180-L1186

And for Python 2.7 I've used the proposed solution from the issue above:
https://github.com/python/mypy/blob/a42ee8f0697439eac3a86644729750c472438b98/mypy/test/teststubgen.py#L147-L153

After that, we've got a green CI.

Sadly, I wasn't able to make a minimal example of the problem since the caching mechanism in Python is not known to me. So anyone interested would need to debug the #6434 (comment out the currently proposed solution for cache invalidation).

I'd welcome any helpful comments on that problem since I'm a bit out of my depth here and not a python expert, but this issue effectively blocks the global hook feature and is a potential sporadic failure for tests/python api in the future.

### PR DESCRIPTION
**Pull request by Minimonium**
Changelog: Bugfix: Fix dynamic imports usage with config installation in the Python API
Docs: Omit

Fixes #6620

### CODE REVIEW

**Comment by Minimonium** [comment]
I'm not sure is it enough to invalidate only on App initialization. 
* In the Python API land - for each API call we create a new App that invalidates the cache. Good.
* We should be wary of a path that creates an App, invalidates the cache, does some import that nulls a finder for a non-existing folder, adds that folder like `conan config install` and _then_ tries to use it. I'm not sure how realistic this scenario is.

**Comment by memsharded** [comment]
I am checking this. It is correct, this import behavior is happening. 
The thing is that it only affects to users of the undocumented python conan_api, otherwise it is not an issue, because a python interpreter is launched for each command. Of course it can also bite us in testing, because test conan commands are evaluated under the same interpreter.

**Comment by Minimonium** [comment]
Thank you for checking it.

**Comment by memsharded** [comment]
Hi @Minimonium 

I have been trying this, understanding the scope, etc. I have pushed a commit https://github.com/conan-io/conan/pull/6633/commits/33262bc6eb7fe8866a9924cb047b41efde4244d4, simplifying a bit the logic. I'd say that a ``sys.path_importer_cache.clear()`` is good enough for python < 3.7. 

Am I missing something?

The other changes are cosmetic, putting the method as private in ConanApp, and minor style to the text.

**Comment by Minimonium** [comment]
I'm not quite familiar with caching myself. The previous <3.7.0 implementation was copied from the fixed version of the importlib in newer versions, but I'm not quite sure how `finder.invalidate_caches()` is important to us.

At least it fixes the issue with the global hook and the implementation is simpler. I'd say that we just should keep in mind that such import situations could happen in case new issues would be reported. Thanks!

**Comment by memsharded** [comment]
Please @jgsogo have a look at the proposal, and to the previous extended implementation by @Minimonium. Thanks!

**Comment by jgsogo** [comment]
From the Conan POV:

The purpose of this PR is to alleviate some people that might be using the `conan_api` in their own Python application, so they can import files that are created dynamically. And the example we are providing uses files inside `<cache>/python` folder, which is marked in the sources like:

```python
        # FIXME Remove in Conan 2.0
        sys.path.append(os.path.join(self.cache_folder, "python"))
```

So we should think twice if the proposed `from myfuncs import mycooladd` is a valid use-case or not to support this feature. 

---

From the Python developer POV:

It is a well know Python behavior that once you start the interpreter, changes in source files are not taken into account (or new modules), and you need to invalidate the cache to achieve that. As we said before these changes are useful for people using the `conan_api` from their own Python application, IMHO I wouldn't expect a library to invalidate the Python cache, it can have unexpected effects on my application (maybe a have a package with plugins I'm downloading and I don't want all them to be loaded).

I, as the developer of the final application, should have enough information about the `conan_api` to decide if invalidating the cache is a good idea or not before or after running some commands.

---

Summing up: we can provide the function, but I really think that we shouldn't call it inside the Conan library. We can document it, and for sure anyone using the `conan_api` should call it after running `conan_api::config_install` if they want to reload the `<cache>/python` folder (**and, more important, the hooks, which probably requires a couple of lines more**).

**Comment by memsharded** [comment]
Thanks for the feedback. I would say:

- Provide some kind of access to that function, expose it in the api (even if not documented, but the idea)
- I would explicitly call that function in the operation that we know will fail if not called: ``conan config install``.

**Comment by Minimonium** [comment]
The invalidating seems to be strongly suggested to be the go-to tool to use per documentation, cpython's tests and various other resources that I read to debug this problem. Alternatively, we can use the part of the extended implementation to just deletes empty finder entries, so no existing imported even if downloaded modules would be affected. We could do the same for Python 2.7 as well. So far I'm not sure I follow since the potential problem sounds very speculative (we can't observe it) and also considering that it would break the observable behaviour of Conan (which we generally don't want to do?).
 
Anyway, I can't claim to understand the problem with invalidating the cache comprehensively enough to argue since I'm not a python expert, so I'll trust you on that one. 

But is it alright that we'd break the behaviour with the global hook then or we're moving it to the 2.0 milestone? Or what's the consensus at all on should we proceed with it or not?

**Comment by memsharded** [comment]
I think we start to have some better understanding of the issues here.

I am not opposed to have the ConanApp clean the caches automatically, that was in fact the original intent of the ConanApp, to have a clean environment for every conan_api call, as we also have some internal caches and not cleaning them in testing between calls was producing errors. This is basically the same problem, so I would be ok with this, the invalidation is a mechanism to deal with some optimization of the interpreter, but it shouldn't really change functionality on the user side (maybe slightly affect performance).

However, this is some base code, a means to an end. The end is that we want to have some kind of "init" hook that is launched every time Conan is called. The most demanded use case by far is the one for calling ``conan config install``. The problem is that it seems very complicated to have some reasonable interface to this, without documenting the api, etc., even with this import issue solved.

I think we need some kind of proof of concept in the line of https://github.com/conan-io/conan/pull/6434 that shows how this functionality will be implemented in a user hook. So I would like to see a test that is doing a ``conan config install`` in the hook (and potentially, to do it only once per day, not every time).

**Comment by Minimonium** [comment]
Gotcha. I'll keep working on it.

**Comment by memsharded** [comment]
We are implementing now the auto "conan config install". As the utility of this is now not that critical, lets implement this properly in the new python api (this is never a problem for the command line) so users have control of the python invalidation if necessary.

**Comment by Minimonium** [comment]
Good to know. Thanks!

**Comment by memsharded** [comment]
Closed by https://github.com/conan-io/conan/pull/10127
